File: analysis/figure_scripts/dalitzfig.py
import matplotlib.pyplot as plt
import sys
import math
from ROOT import TChain
from itertools import permutations
from dataframe import *

chain = TChain('a')
for f in input().split(): chain.AddFile(f)
fname = sys.argv[1]+"/"+sys.argv[0].split('/')[2].split('.')[0]

X = []
Y = []
for i,j,k in list(permutations([0,1,2])):
    X.append("((eCM["+str(j)+"]-eCM["+str(k)+"])/esum)*sqrt(3)")
    Y.append("3*eCM["+str(i)+"]/esum-1.0")
r = dataframe(chain)
r = r.define("esum","eCM[0]+eCM[1]+eCM[2]")

xlim = (-1.3,1.3)
ylim = (-1.3,1.3)

#plt.figure(figsize=(7*7/10,4.5*7/10))
width =  1.0/72.27*393
plt.figure(figsize=(width/2,width/1.61803398875*0.7))
##plt.title()
r = r.filter("mul==3")\
    .filter("pT<50e3")\
    .filter("abs(deltaE)<200")\
    .define("E2","min(eCM[0],min(eCM[1],eCM[2]))")\
    .define("E1","max(min(eCM[0],eCM[1]), min(max(eCM[0],eCM[1]),eCM[2]))")\
    .define("E0","max(eCM[0],max(eCM[1],eCM[2]))")\
    .define("X","((E1-E2)/esum)*sqrt(3)")\
    .define("Y","3*E0/esum-1.0")\
    .filter("pow(X,2)+pow(Y,2)<1.3")\
    .filter("E2>50")
    #.filter("Y<0.93")\

ghist2d(r,X,Y,200,(xlim,ylim),fname)
#if "wU" in r.getColumnNames():
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"f1"),weights=["wU*f[0][0]"]*6)
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"f2"),weights=["wU*f[0][1]"]*6)
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"im"),weights=["wU*f[0][3]"]*6)
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"re"),weights=["wU*f[0][2]"]*6)

#if "data" in fname:
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"count"))
    #print("data")
#else:
    #print("not data")
    #r\
    #.hist2d(X,Y,bins=200,range=(xlim,ylim),export=(fname,"count"),weights=["w0*w1*w2"]*6)

#plt.clf()
#if "wU" in r.getColumnNames():
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),weights=["wU*(0.22*f[0][0]+(1-0.22)*f[0][1]+2*sqrt(0.22*(1-0.22))*(f[0][2]*cos(2*3.1415*0.69)+f[0][3]*sin(2*3.1415*0.69)))*w0*w1*w2"]*6)
#elif "data" in fname:
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim))
#else:
    #r.hist2d(X,Y,bins=200,range=(xlim,ylim),weights=["w0*w1*w2"]*6)


#r.hist2d(X,Y,bins=250,range=(xlim,ylim),export=fname)
plt.xlabel("$X$")
plt.ylabel("$Y$")
plt.gca().zaxis.set_scale('log')
plt.gca().set_xlim(*xlim)
plt.gca().set_ylim(*ylim)
plt.gca().set_aspect("equal")

plt.savefig(fname+".pdf",dpi=1200)

File: analysis/figure_scripts/dalitzfig.cpp
// ROOT stuff
#include <TFile.h>
#include <TTree.h>
#include <TChain.h>
#include <ROOT/RDataFrame.hxx>
#include <ROOT/RDFHelpers.hxx>
#include <TStyle.h>
#include <TROOT.h>
#include <TCanvas.h>

// other stuff
#include <filesystem>
#include <boost/format.hpp>
#include <iostream>

// my stuff
#include "../plot_style.cpp"

using namespace std;
using namespace ROOT;
using boost::format;

int main(int argc, char *argv[]) {
    // all but the last argument are input files
    TChain chain("tree");
    for (int i = 1; i < argc-1; i++) {
        chain.Add(argv[i]);
    }    
    // final argument is the destination
    string dest = argv[argc-1]; 
    dest += "dalitzfig.pdf";

    // prepare the dataframe
    ROOT::RDF::RNode df = RDataFrame(chain);
    df = df.Define("E_sum","E_cm[0]+E_cm[1]+E_cm[2]");

    // set the axes    
    double x_axis[] = {200, -1.3, 1.3};
    double y_axis[] = {200, -1.3, 1.3};

    // define the necessary variables
    df = df.Filter("p_tot<35e3")
        .Filter("abs(deltaE)<200")
        .Define("E2","min(E_cm[0],min(E_cm[1],E_cm[2]))")
        .Define("E1","max(min(E_cm[0],E_cm[1]), min(max(E_cm[0],E_cm[1]),E_cm[2]))")
        .Define("E0","max(E_cm[0],max(E_cm[1],E_cm[2]))")
        .Define("X","((E1-E2)/E_sum)*sqrt(3)")
        .Define("Y","3*E0/E_sum-1.0")
        .Filter("pow(X,2)+pow(Y,2)<1.0")
//        .Filter("Y<0.93")
        .Filter("E2>250");

    //*** PLOT ***//
    setup_style();

    TCanvas* canvas = new TCanvas("Dalitz_fig", "Dalitz_fig", 600, 600);
    TH2D* hist = new TH2D("h1", "kinematic_lab", int(x_axis[0]), x_axis[1], x_axis[2], int(y_axis[0]), y_axis[1], y_axis[2]);

    int perms[] = {0, 1, 2};
    do {
        int i = perms[0];
        int j = perms[1];
        int k = perms[2];
        TH2D htemp = df.Define("x", (format("((E_cm[%1%] - E_cm[%2%])/E_sum)*sqrt(3)") % j % k).str())
                       .Define("y", (format("3*E_cm[%1%]/E_sum-1") % i).str())
                       .Histo2D({"h1", "kinematic_lab", int(x_axis[0]), x_axis[1], x_axis[2], int(y_axis[0]), y_axis[1], y_axis[2]}, "x", "y").GetValue();
        hist->Add(&htemp);
    } while (std::next_permutation(perms, perms+3));
    
    hist->GetXaxis()->SetTitle("X");
    hist->GetYaxis()->SetTitle("Y");
    hist->Draw("colz");

    canvas->SetLogz();
    canvas->SetRightMargin(0.15);
    canvas->SaveAs(dest.c_str());
}